### ISSUE TITLE
WebAuthN registration broken after upgrading to 24.0.1

### ISSUE THREAD

**Issue by jlrcontegix** [keycloak/keycloak#28113]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

authentication/webauthn

### Describe the bug

We had the following settings configured for our WebAuthN policy in 23.0.6:

Attention conveyance preference: Direct
Authenticator Attachment: Cross Platform
Require discoverable credential: Yes
Acceptable AAGUIDs: c1f9a0bc-1dd2-404a-b27f-8e29047a43fd

After upgrading to 24.0.1 users began reporting issues registering their Yubikeys. When I look at the admin panel in 24.0.1, The "Authenticator Attachment" setting has been changed to "Platform". When selecting "Cross Platform" it reverts back to "Platform". I had to change the setting to Not Specified in order for it work with an external USB key.

When users attempt to register a Yubikey with the correct AAGUID they are presented with the following error:

Failed to register your Passkey.
not acceptable aaguid = 00000000-0000-0000-0000-000000000000

This was working fine yesterday when we were on 23.0.6. In the logs we see the following when a user attempts to register their key:

```
Mar 20 16:14:58 server kc.sh[999205]: 2024-03-20 16:14:58,567 WARN  [org.keycloak.authentication.requiredactions.WebAuthnRegister] (executor-thread-5) webauthn-error-registration
Mar 20 16:14:58 server kc.sh[999205]: 2024-03-20 16:14:58,576 WARN  [org.keycloak.events] (executor-thread-5) type="CUSTOM_REQUIRED_ACTION_ERROR", realmId="XXXXX", clientId="account-console", userId="XXXXX", ipAddress="XXXXX", error="invalid_registration", credential_type="webauthn", auth_method="openid-connect", web_authn_registration_error_detail="not acceptable aaguid = 00000000-0000-0000-0000-000000000000", custom_required_action="webauthn-register", response_type="code", web_authn_registration_error="webauthn-error-registration", redirect_uri="XXXXX", remember_me="false", code_id="5700685b-ea97-474c-9998-ede4820bd15f", response_mode="query", username="XXXXX"
```

If I remove the AAGUID from the admin panel the users are able to register the key, but we cannot allow any AAGUID, so this is a big issue for us.

### Version

24.0.1

### Regression

- [X] The issue is a regression

### Expected behavior

Users should be able to register their Yubikey given the settings that worked in 23.0.6.

### Actual behavior

Users are presented with the following error unless we remove the AAGUID from the admin panel:
Failed to register your Passkey.
not acceptable aaguid = 00000000-0000-0000-0000-000000000000

### How to Reproduce?

Login as a user without a WebAuthN credential attached and go through the self registration process.

**Comment by jlrcontegix** [comment]
Further testing confirms we have to set all of our previous values to "Not Specified" and remove the AAGUID in order for the registration complete. If try to specify any of the following the registration fails with varying errors:

Attention conveyance preference
Authenticator Attachment
Require discoverable credential
Acceptable AAGUIDs

**Comment by rmartinc** [comment]
Hi @jlrcontegix!

We have another issues with webauthn so probably there is something wrong in 24.0.1 (for example https://github.com/keycloak/keycloak/issues/27890). I have checked and the `cross-platform` issue, and it is just something wrong with the admin console. In my tests if I set `Authentication Attachment` to `Cross Platform`, although it's seen as `Platform` in the console it really saves `cross-platform` the the realm.

You can check (and set) what you have using for example `kcadm.sh` (I'm checking setting both, normal and passwordless):

```
./kcadm.sh get -r <realm_name> / | grep -i attachment
  "webAuthnPolicyAuthenticatorAttachment" : "cross-platform",
  "webAuthnPolicyPasswordlessAuthenticatorAttachment" : "cross-platform",
./kcadm.sh update -r test / -s "webAuthnPolicyPasswordlessAuthenticatorAttachment=cross-platform" -s "webAuthnPolicyAuthenticatorAttachment=cross-platform"
```

If you set to cross-platform like in 23, is it working again? (I think there is something more than the cross-platform issue in the console.)

**Comment by rmartinc** [comment]
I have created https://github.com/keycloak/keycloak/issues/28127 for the display issue. UI team will check it if my weird solution is OK or they prefer something different.

**Comment by jlrcontegix** [comment]
You are correct, it does indeed set the value to Cross Platform under the hood, but as you stated there are still other issues. 

When I set the following values but leave the AAGUID empty we see a different error:

Attestation conveyance preference: Direct
Authenticator Attachment: Cross Platform

Error displayed in browser:

Failed to register your Passkey.
invalid cert path

Error in logs:
```
Mar 21 07:58:00 server kc.sh[999205]: 2024-03-21 07:58:00,013 WARN  [org.keycloak.events] (executor-thread-245) type="CUSTOM_REQUIRED_ACTION_ERROR", realmId="XXXXX", clientId="account-console", userId="XXXXX", ipAddress="XXXXX", error="invalid_registration", credential_type="webauthn", auth_method="openid-connect", web_authn_registration_error_detail="invalid cert path", custom_required_action="webauthn-register", response_type="code", web_authn_registration_error="webauthn-error-registration", redirect_uri="https://sso.contegix.com/realms/contegix/account/", remember_me="false", code_id="XXXXX", response_mode="query", username="XXXXX"
```

**Comment by rmartinc** [comment]
@jlrcontegix My feeling is the issue you are seeing is because of the changes in the truststore provider in 24 (issue https://github.com/keycloak/keycloak/issues/24148). Did you configure previously the truststore provider with a file? Or are you starting the server with no truststore?

Maybe the problem is that previously there were no truststore and the [NullCertPathTrustworthinessValidator](https://github.com/keycloak/keycloak/blob/24.0.1/services/src/main/java/org/keycloak/authentication/requiredactions/WebAuthnRegisterFactory.java#L43) and now you are receiving the java cacerts and the certificate sent by the attestation is not valid for them. Does it make sense to you?

If that is the case, I would change the attestation to None and I try to configure the truststore with the root certificate that the yubikey is using.

**Comment by jlrcontegix** [comment]
@rmartinc We are not starting the server with a truststore and hadn't previously configured one.

I just changed the Attestation to None and added Yubico's CA to conf/truststores but still receive this error:
Failed to register your Passkey.
not acceptable aaguid = 00000000-0000-0000-0000-000000000000

**Comment by rmartinc** [comment]
But it's that AAGUID added as valid?

**Comment by jlrcontegix** [comment]
The Yubikey I am registering has AAGUID c1f9a0bc-1dd2-404a-b27f-8e29047a43fd, which is what we've set in the Keycloak interface under Acceptable AAGUIDs. The error includes all zeros as shown but that's obviously not the AAGUID of the device or what is specified in the admin UI.

**Comment by jlrcontegix** [comment]
Did some more testing and it does work if I change Attestation conveyance preference to Indirect.

**Comment by rmartinc** [comment]
Ah! yes, I think that the WebAuthn spec defines that when "none" attestation is requested, client must replace the AAGUID with 16 zero bytes. Some clients ignores this but yours seem to be OK. So you have to add hat aaguid for the moment.

**Comment by rmartinc** [comment]
Thanks @jlrcontegix! This is a change in behavior in the webauthn but IMHO now it is more secure and the way it should have been working before. Previously if you didn't define any truststore any certificate sent by the webauthn request was valid using the NullCertPathTrustworthinessValidator. Now, after the changes in the truststore for 24, the NullCertPathTrustworthinessValidator is never used because the truststore always incorporates the java certificates. This was not something done on purpose, just collateral effect of the truststore change.

In summary, now if webauthn configuration requires attestation conveyance (Direct for example) the certificate should always be provided via truststore.

**Comment by jlrcontegix** [comment]
I appreciate you @rmartinc! I would have stumbled on this for a while if not for the pointer about the truststore. I read the upgrade notes about the change prior to upgrading, but since we've never used the truststore, I assumed it didn't apply to us. This issue might serve as help to others who may have had similar configuration prior and have trouble after upgrading to 24.x. Thanks for the quick help.

**Comment by rmartinc** [comment]
We have decided to add a paragraph in the trust-store upgrading notes about webauthn registration. So re-open this to use the issue for the PR.

**Comment by JCHEG** [comment]
> Okay, I think we can close this, after changing Attestation conveyance preference back to Direct it works, so I believe the root of my issue was I needed the Yubico CA in the truststore.

Hello,
Please, could you tell me where you found the Youbico certificate and how you installes it ?
I tried 4 differents certificates putting them in ```/opt/keycloak/conf/truststores``` but it doesn't work, i get ```invalid cert path``` 
In the console logs i have 
```SHELL
2024-08-27 10:16:56,127 WARN  [org.keycloak.authentication.requiredactions.WebAuthnRegister] (executor-thread-9) webauthn-error-registration
2024-08-27 10:16:56,128 WARN  [org.keycloak.events] (executor-thread-9) type="CUSTOM_REQUIRED_ACTION_ERROR", realmId="XXX", realmName="XXX", clientId="helper-portal-backoffice-v1", userId="XXX", ipAddress="XXX", error="invalid_registration", credential_type="webauthn", auth_method="openid-connect", web_authn_registration_error_detail="invalid cert path", custom_required_action="webauthn-register", response_type="code", web_authn_registration_error="webauthn-error-registration", redirect_uri="https://XXX/index.php", remember_me="false", code_id="XXX", response_mode="query", username="XXX"
```

**Comment by JCHEG** [comment]
I found this which worked for me putting these 2 yubico certificates in keycloak trustore:

[https://developers.yubico.com/FIDO/yubico-fido-ca-certs.txt](url)
